File: exercises/01_hello_tailnet/solution/workflow.py
# ABOUTME: Temporal Workflow that gets your public IP and geolocates it.
# ABOUTME: Demonstrates basic workflow structure with two sequential activities.

import asyncio
from datetime import timedelta

from shared import WorkflowInput, WorkflowOutput
from temporalio import workflow

with workflow.unsafe.imports_passed_through():
    from activities import get_ip, get_location_info


@workflow.defn
class GetAddressFromIP:
    @workflow.run
    async def run(self, input: WorkflowInput) -> WorkflowOutput:
        ip_address = await workflow.execute_activity(
            get_ip,
            start_to_close_timeout=timedelta(seconds=5),
        )

        await asyncio.sleep(input.seconds)

        location = await workflow.execute_activity(
            get_location_info,
            ip_address,
            start_to_close_timeout=timedelta(seconds=5),
        )

        return WorkflowOutput(ip_addr=ip_address, location=location)

File: exercises/02_explore_tailscale/go-hello-tsnet/solution/workflow.go
package main

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/workflow"
)

func GetAddressFromIP(ctx workflow.Context, input WorkflowInput) (WorkflowOutput, error) {
	ao := workflow.ActivityOptions{
		StartToCloseTimeout: time.Minute,
	}
	ctx = workflow.WithActivityOptions(ctx, ao)

	var ip string
	if err := workflow.ExecuteActivity(ctx, GetIP).Get(ctx, &ip); err != nil {
		return WorkflowOutput{}, fmt.Errorf("get ip: %w", err)
	}

	var location string
	if err := workflow.ExecuteActivity(ctx, GetLocationInfo, ip).Get(ctx, &location); err != nil {
		return WorkflowOutput{}, fmt.Errorf("get location: %w", err)
	}

	return WorkflowOutput{IPAddr: ip, Location: location}, nil
}
